File: leetcode/go/linkedList/removeNthNodeFromEndOfList(medium)/removeNthNodeFromEndOfList.go
package main

import "fmt"

/* 19. Remove Nth Node From End of List
https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/

Дан заголовок связанного списка, удалить n-й узел из конца списка и вернуть его заголовок.

Example 1:
Input: head = [1,2,3,4,5], n = 2
Output: [1,2,3,5]

Example 2:
Input: head = [1], n = 1
Output: []
*/

func main() {
	head1 := createList([]int{1, 2, 3, 4, 5})
	n := 2
	result1 := removeNthFromEnd(head1, n)
	printList(result1) // [1, 2, 3, 5]

	head2 := createList([]int{1, 2, 3, 4, 5})
	result2 := removeNthFromEndVariantSlowFast(head2, n)
	printList(result2) // [1, 2, 3, 5]
}

type ListNode struct {
	Val  int
	Next *ListNode
}

// removeNthFromEnd удаляет n-й узел из конца списка и возвращает его заголовок
// Паттерн Dummy
// time complexity: O(n), где n - количество узлов в списке
// space complexity: O(1)
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	// Создаем фиктивный узел dummyNode, чтобы избежать специальных случаев с пустым списком или удалением первого узла.
	dummyNode := &ListNode{Val: 0, Next: head}

	// Вычисляем длину списка с учетом dummyNode
	length := 0
	curr := dummyNode
	for curr != nil {
		curr = curr.Next
		length++
	}

	// Доходим до (n-1)-ой вершины с конца
	curr = dummyNode
	// Тут условие i < length-n-1, потому что нам нужно дойти до (n-1)-ой вершины с конца, а не до n-ой вершины.
	for i := 0; i < length-n-1; i++ {
		curr = curr.Next // Двигаемся к следующему узлу
	}

	// Удаляем вершину
	// Проверяем, что узел существует, прежде чем удалить его
	if curr.Next != nil {
		// Удаляем вершину, изменяя ссылку на следующую вершину
		curr.Next = curr.Next.Next
	}

	// Возвращаем заголовок списка, начиная с dummyNode
	return dummyNode.Next
}

// removeNthFromEndVariantSlowFast удаляет n-й узел из конца списка и возвращает его заголовок
// time complexity: O(n), где n - количество узлов в списке
// space complexity: O(1)
func removeNthFromEndVariantSlowFast(head *ListNode, n int) *ListNode {
	// Создаем фиктивный узел dummyNode, чтобы избежать специальных случаев с пустым списком или удалением первого узла.
	dummyNode := &ListNode{Val: 0, Next: head}

	fast := dummyNode
	// Перемещаем быстрый указатель на n+1 позиций вперед
	for i := 0; i <= n; i++ {
		fast = fast.Next
	}

	slow := dummyNode
	// Перемещаем оба указателя, пока быстрый не достигнет конца
	for fast != nil {
		slow = slow.Next
		fast = fast.Next
	}

	// Удаляем n-й узел с конца
	slow.Next = slow.Next.Next

	// Возвращаем заголовок списка, начиная с dummyNode
	return dummyNode.Next
}

// createList создает связный список из массива значений
func createList(values []int) *ListNode {
	if len(values) == 0 {
		return nil
	}

	head := &ListNode{Val: values[0]}
	current := head

	for _, val := range values[1:] {
		current.Next = &ListNode{Val: val}
		current = current.Next
	}

	return head
}

// printList печатает список в формате [v1,v2,v3]
func printList(head *ListNode) {
	fmt.Print("[")
	for head != nil {
		fmt.Print(head.Val)
		if head.Next != nil {
			fmt.Print(",")
		}
		head = head.Next
	}
	fmt.Println("]")
}

File: leetcode/java/linkedList/RemoveNthNodeFromEndOfList/RemoveNthNodeFromEndOfList.java
package leetcode.java.linkedList.RemoveNthNodeFromEndOfList;

/* 19. Remove Nth Node From End of List
https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/

Дан заголовок связанного списка, удалить n-й узел из конца списка и вернуть его заголовок.

Example 1:
Input: head = [1,2,3,4,5], n = 2
Output: [1,2,3,5]

Example 2:
Input: head = [1], n = 1
Output: []
*/

public class RemoveNthNodeFromEndOfList {
    public static void main(String[] args) {
        ListNode head1 = createList(new int[]{1, 2, 3, 4, 5});
        int n = 2;
        ListNode result1 = removeNthFromEnd(head1, n);
        printList(result1); // [1, 2, 3, 5]

        ListNode head2 = createList(new int[]{1, 2, 3, 4, 5});
        ListNode result2 = removeNthFromEndVariantSlowFast(head2, n);
        printList(result2); // [1, 2, 3, 5]
    }

    static class ListNode {
        int val;
        ListNode next;

        ListNode(int val) {
            this.val = val;
        }
    }

    /**
     * removeNthFromEnd удаляет n-й узел из конца списка и возвращает его заголовок.
     * Паттерн Dummy + вычисление длины списка.
     * time: O(n), где n - количество узлов в списке
     * space: O(1)
     */
    private static ListNode removeNthFromEnd(ListNode head, int n) {
        // Создаем фиктивный узел dummyNode, чтобы избежать специальных случаев
        // с пустым списком или удалением первого узла
        ListNode dummyNode = new ListNode(0);
        dummyNode.next = head;

        // Вычисляем длину списка с учетом dummyNode
        int length = 0;
        ListNode curr = dummyNode;
        while (curr != null) {
            curr = curr.next;
            length++;
        }

        // Доходим до (n-1)-ой вершины с конца
        curr = dummyNode;
        // Условие i < length-n-1, потому что нам нужно дойти до узла ПЕРЕД удаляемым
        for (int i = 0; i < length - n - 1; i++) {
            curr = curr.next;
        }

        // Удаляем вершину
        // Проверяем, что узел существует, прежде чем удалить его
        if (curr.next != null) {
            curr.next = curr.next.next;
        }

        // Возвращаем заголовок списка, начиная с dummyNode
        return dummyNode.next;
    }

    /**
     * removeNthFromEndVariantSlowFast удаляет n-й узел из конца списка.
     * Паттерн Two Pointers (slow/fast) + Dummy.
     * time: O(n), где n - количество узлов в списке
     * space: O(1)
     */
    private static ListNode removeNthFromEndVariantSlowFast(ListNode head, int n) {
        // Создаем фиктивный узел dummyNode для обработки краевых случаев
        ListNode dummyNode = new ListNode(0);
        dummyNode.next = head;

        ListNode fast = dummyNode;
        // Перемещаем быстрый указатель на n+1 позиций вперед
        for (int i = 0; i <= n; i++) {
            fast = fast.next;
        }

        ListNode slow = dummyNode;
        // Перемещаем оба указателя, пока быстрый не достигнет конца
        while (fast != null) {
            slow = slow.next;
            fast = fast.next;
        }

        // Удаляем n-й узел с конца
        if (slow.next != null) {
            slow.next = slow.next.next;
        }

        // Возвращаем заголовок списка
        return dummyNode.next;
    }

    /**
     * createList создает связный список из массива значений.
     */
    private static ListNode createList(int[] values) {
        if (values == null || values.length == 0) {
            return null;
        }

        ListNode head = new ListNode(values[0]);
        ListNode current = head;

        for (int i = 1; i < values.length; i++) {
            current.next = new ListNode(values[i]);
            current = current.next;
        }

        return head;
    }

    /**
     * printList печатает список в формате [v1,v2,v3]
     */
    private static void printList(ListNode head) {
        System.out.print("[");
        while (head != null) {
            System.out.print(head.val);
            if (head.next != null) {
                System.out.print(",");
            }
            head = head.next;
        }
        System.out.println("]");
    }
}
